### PR TITLE
MS-440 Create kino microphone input

### PR DESCRIPTION
**Pull request by Janix4000**

### CODE REVIEW

**Comment by mat-hek** [comment]
@Janix4000 I get the following error when I try to run the audio and video example:
```

11:09:57.776 [error] GenServer #PID<0.4385.0> terminating
** (Membrane.StreamFormatError) Stream format: %Membrane.H264{width: 1920, height: 1080, profile: :high, alignment: :au, nalu_in_metadata?: false, framerate: {60, 1}} is not matching accepted format pattern in def_input_pad
for pad :video in Membrane.Kino.Player.Sink

    (membrane_core 0.11.3) lib/membrane/core/element/stream_format_controller.ex:96: anonymous fn/4 in Membrane.Core.Element.StreamFormatController.validate_stream_format!/3
    (elixir 1.14.2) lib/enum.ex:2468: Enum."-reduce/3-lists^foldl/2-0-"/3
    (membrane_core 0.11.3) lib/membrane/core/element/stream_format_controller.ex:94: Membrane.Core.Element.StreamFormatController.validate_stream_format!/3
    (membrane_core 0.11.3) lib/membrane/core/element/stream_format_controller.ex:64: Membrane.Core.Element.StreamFormatController.exec_handle_stream_format/4
    (membrane_core 0.11.3) lib/membrane/core/element.ex:183: Membrane.Core.Element.handle_info/2
    (stdlib 4.0.1) gen_server.erl:1120: :gen_server.try_dispatch/4
    (stdlib 4.0.1) gen_server.erl:1197: :gen_server.handle_msg/6
    (stdlib 4.0.1) proc_lib.erl:240: :proc_lib.init_p_do_apply/3
Last message: {Membrane.Core.Message, :stream_format, %Membrane.H264{width: 1920, height: 1080, profile: :high, alignment: :au, nalu_in_metadata?: false, framerate: {60, 1}}, [for_pad: {Membrane.Pad, :video, #Reference<0.1596432836.420478978.18292>}]}
```

**Comment by mat-hek** [comment]
The audio demo says it uses a video player 🤔 and it sometimes crashes with:
```
11:15:06.252 [error] GenServer #PID<0.460.0> terminating
** (stop) exited in: GenServer.call(#PID<0.339.0>, {:create, 43.06640625}, 5000)
    ** (EXIT) time out
    (elixir 1.14.2) lib/gen_server.ex:1038: GenServer.call/3
    (membrane_kino_plugin 0.1.0) lib/membrane/kino/player_sink.ex:183: Membrane.Kino.Player.Sink.create_player/2
    (membrane_kino_plugin 0.1.0) lib/membrane/kino/player_sink.ex:165: Membrane.Kino.Player.Sink.handle_start_of_stream/3
    (membrane_core 0.11.3) lib/membrane/core/callback_handler.ex:138: Membrane.Core.CallbackHandler.exec_callback/4
    (membrane_core 0.11.3) lib/membrane/core/callback_handler.ex:62: Membrane.Core.CallbackHandler.exec_and_handle_callback/5
    (membrane_core 0.11.3) lib/membrane/core/element/event_controller.ex:79: Membrane.Core.Element.EventController.do_exec_handle_event/4
    (membrane_core 0.11.3) lib/membrane/core/element/buffer_controller.ex:43: Membrane.Core.Element.BufferController.handle_buffer/3
    (membrane_core 0.11.3) lib/membrane/core/element.ex:177: Membrane.Core.Element.handle_info/2
Last message: {Membrane.Core.Message, :buffer, [%Membrane.Buffer{payload: <<255, 241, 80, 64, 22, 127, 252, 222, 2, 0, 76, 97, 118, 99, 53, 56, 46, 53, 52, 46, 49, 48, 48, 0, 2, 116, 171, 84, 216, 228, 109, 89, 85, 223, 63, 90, 185, 89, 89, 239, 223, 227, 87, 34, 32, ...>>, pts: 0, dts: nil, metadata: %{}}, %Membrane.Buffer{payload: <<255, 241, 80, 64, 32, 159, 252, 1, 40, 152, 218, 224, 173, 215, 9, 140, 3, 110, 146, 97, 55, 35, 34, 1, 221, 196, 226, 236, 28, 121, 154, 212, 254, 253, 119, 215, 71, 183, 28, 76, 255, 226, 87, 158, ...>>, pts: 10240000000 <|> 441, dts: nil, metadata: %{}}, %Membrane.Buffer{payload: <<255, 241, 80, 64, 26, 159, 252, 1, 18, 245, 172, 208, 139, 27, 16, 172, 255, 176, 22, 192, 1, 17, 4, 16, 0, 93, 68, 160, 136, 133, 42, 84, 46, 218, 119, 72, 212, 44, 32, 53, 216, 88, 213, ...>>, pts: 20480000000 <|> 441, dts: nil, metadata: %{}}, %Membrane.Buffer{payload: <<255, 241, 80, 64, 23, 191, 252, 0, 230, 53, 172, 236, 244, 26, 152, 88, 241, 159, 145, 237, 174, 26, 33, 42, 42, 12, 82, 148, 20, 170, 80, 166, 170, 149, 103, 10, 178, 168, 60, 144, 146, 114, ...>>, pts: 10240000000 <|> 147, dts: nil, metadata: %{}}, %Membrane.Buffer{payload: <<255, 241, 80, 64, 27, 127, 252, 0, 228, 53, 172, 209, 6, 34, 12, 138, 130, 19, 189, 63, 52, 124, 107, 83, 66, 0, 21, 74, 80, 82, 148, 161, 136, 229, 39, 43, 153, 202, 164, 98, 149, ...>>, pts: 40960000000 <|> 441, dts: nil, metadata: %{}}, %Membrane.Buffer{payload: <<255, 241, 80, 64, 24, 95, 252, 0, 234, 53, 172, 236, 231, 10, 13, 78, 43, 120, 207, 145, 237, 122, 156, 9, 25, 106, 133, 82, 148, 160, 165, 82, 132, 243, 77, 44, 173, 44, 171, 2, ...>>, pts: 51200000000 <|> 441, dts: nil, metadata: %{}}, %Membrane.Buffer{payload: <<255, 241, 80, 64, 28, 31, 252, 0, 228, 53, 172, 244, 70, 60, 17, 136, 137, 34, 9, 28, 191, 129, 241, 174, 36, 64, 0, 82, 148, 81, 74, 80, 59, 164, 86, 228, 46, 0, 231, ...>>, pts: 20480000000 <|> 147, dts: nil, metadata: %{}}, %Membrane.Buffer{payload: <<255, 241, 80, 64, 23, 31, 252, 0, 238, 53, 172, 236, 244, 56, 185, 203, 244, 31, 23, 195, 80, 68, 50, 202, 170, 82, 130, 148, 165, 9, 234, 121, 106, 105, 106, 89, 4, 228, ...>>, pts: 10240000000 <|> 63, dts: nil, metadata: %{}}, %Membrane.Buffer{payload: <<255, 241, 80, 64, 28, 127, 252, 0, 228, 53, 172, 240, 116, 35, 133, 146, 135, 34, 9, 28, 191, 129, 241, 174, 46, 0, 0, 82, 148, 41, 74, 80, 62, 224, 91, 104, 135, ...>>, pts: 81920000000 <|> 441, dts: nil, metadata: %{}}], [for_pad: {Membrane.Pad, :audio, #Reference<0.3252172004.3910402053.66272>}]}
```

**Comment by mat-hek** [comment]
And I don't see examples for the sources

**Review by mat-hek — COMMENTED** [review]
see the comments above

**Comment by mat-hek** [comment]
There's a race condition triggered by evaluating the last cell without running the previous ones (they're triggered automatically) causing the following error:

```
14:50:49.110 [error] GenServer #PID<0.3213.0> terminating
** (stop) exited in: GenServer.call(#PID<0.3199.0>, {:create, 43.06640625}, 5000)
    ** (EXIT) time out
    (elixir 1.14.2) lib/gen_server.ex:1038: GenServer.call/3
    (membrane_kino_plugin 0.1.0) lib/membrane/kino/player_sink.ex:188: Membrane.Kino.Player.Sink.create_player/2
    (membrane_kino_plugin 0.1.0) lib/membrane/kino/player_sink.ex:169: Membrane.Kino.Player.Sink.handle_start_of_stream/3
    (membrane_core 0.11.3) lib/membrane/core/callback_handler.ex:138: Membrane.Core.CallbackHandler.exec_callback/4
    (membrane_core 0.11.3) lib/membrane/core/callback_handler.ex:62: Membrane.Core.CallbackHandler.exec_and_handle_callback/5
    (membrane_core 0.11.3) lib/membrane/core/element/event_controller.ex:79: Membrane.Core.Element.EventController.do_exec_handle_event/4
    (membrane_core 0.11.3) lib/membrane/core/element/buffer_controller.ex:43: Membrane.Core.Element.BufferController.handle_buffer/3
    (membrane_core 0.11.3) lib/membrane/core/element.ex:177: Membrane.Core.Element.handle_info/2
Last message: {Membrane.Core.Message, :buffer, [%Membrane.Buffer{payload: <<255, 241, 80, 64, 22, 127, 252, 222, 2, 0, 76, 97, 118, 99, 53, 56, 46, 53, 52, 46, 49, 48, 48, 0, 2, 116, 171, 84, 216, 228, 109, 89, 85, 223, 63, 90, 185, 89, 89, 239, 223, 227, 87, 34, 32, ...>>, pts: 0, dts: nil, metadata: %{}}, %Membrane.Buffer{payload: <<255, 241, 80, 64, 32, 159, 252, 1, 40, 152, 218, 224, 173, 215, 9, 140, 3, 110, 146, 97, 55, 35, 34, 1, 221, 196, 226, 236, 28, 121, 154, 212, 254, 253, 119, 215, 71, 183, 28, 76, 255, 226, 87, 158, ...>>, pts: 10240000000 <|> 441, dts: nil, metadata: %{}}, %Membrane.Buffer{payload: <<255, 241, 80, 64, 26, 159, 252, 1, 18, 245, 172, 208, 139, 27, 16, 172, 255, 176, 22, 192, 1, 17, 4, 16, 0, 93, 68, 160, 136, 133, 42, 84, 46, 218, 119, 72, 212, 44, 32, 53, 216, 88, 213, ...>>, pts: 20480000000 <|> 441, dts: nil, metadata: %{}}, %Membrane.Buffer{payload: <<255, 241, 80, 64, 23, 191, 252, 0, 230, 53, 172, 236, 244, 26, 152, 88, 241, 159, 145, 237, 174, 26, 33, 42, 42, 12, 82, 148, 20, 170, 80, 166, 170, 149, 103, 10, 178, 168, 60, 144, 146, 114, ...>>, pts: 10240000000 <|> 147, dts: nil, metadata: %{}}, %Membrane.Buffer{payload: <<255, 241, 80, 64, 27, 127, 252, 0, 228, 53, 172, 209, 6, 34, 12, 138, 130, 19, 189, 63, 52, 124, 107, 83, 66, 0, 21, 74, 80, 82, 148, 161, 136, 229, 39, 43, 153, 202, 164, 98, 149, ...>>, pts: 40960000000 <|> 441, dts: nil, metadata: %{}}, %Membrane.Buffer{payload: <<255, 241, 80, 64, 24, 95, 252, 0, 234, 53, 172, 236, 231, 10, 13, 78, 43, 120, 207, 145, 237, 122, 156, 9, 25, 106, 133, 82, 148, 160, 165, 82, 132, 243, 77, 44, 173, 44, 171, 2, ...>>, pts: 51200000000 <|> 441, dts: nil, metadata: %{}}, %Membrane.Buffer{payload: <<255, 241, 80, 64, 28, 31, 252, 0, 228, 53, 172, 244, 70, 60, 17, 136, 137, 34, 9, 28, 191, 129, 241, 174, 36, 64, 0, 82, 148, 81, 74, 80, 59, 164, 86, 228, 46, 0, 231, ...>>, pts: 20480000000 <|> 147, dts: nil, metadata: %{}}, %Membrane.Buffer{payload: <<255, 241, 80, 64, 23, 31, 252, 0, 238, 53, 172, 236, 244, 56, 185, 203, 244, 31, 23, 195, 80, 68, 50, 202, 170, 82, 130, 148, 165, 9, 234, 121, 106, 105, 106, 89, 4, 228, ...>>, pts: 10240000000 <|> 63, dts: nil, metadata: %{}}, %Membrane.Buffer{payload: <<255, 241, 80, 64, 28, 127, 252, 0, 228, 53, 172, 240, 116, 35, 133, 146, 135, 34, 9, 28, 191, 129, 241, 174, 46, 0, 0, 82, 148, 41, 74, 80, 62, 224, 91, 104, 135, ...>>, pts: 81920000000 <|> 441, dts: nil, metadata: %{}}], [for_pad: {Membrane.Pad, :audio, #Reference<0.2141576257.800325637.66363>}]}
```

This error seems to happen when kino is occupied, a moment after `Kino source already occupied` error.

File: lib/membrane/kino/input.ex [hunk 1]
```diff
@@ -0,0 +1,81 @@
+defmodule Membrane.Kino.Input do
+  defmodule InputError do
+    defexception [:message]
+  end
+
+  use Kino.JS, assets_path: "lib/assets/audio_input"
```
Comment (mat-hek) [inline lib/membrane/kino/input.ex:6]: I think that assets should live in `priv`

File: lib/membrane/kino/input.ex [hunk 2]
```diff
@@ -0,0 +1,81 @@
+defmodule Membrane.Kino.Input do
```
Comment (mat-hek) [inline lib/membrane/kino/input.ex:1]: no moduledoc

File: lib/membrane/kino/input.ex [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+  defmodule InputError do
+    defexception [:message]
+  end
+
+  use Kino.JS, assets_path: "lib/assets/audio_input"
+  use Kino.JS.Live
+
+  alias Membrane.Time
+
+  def new(_type \\ :audio, opts \\ []) do
```
Comment (mat-hek) [inline lib/membrane/kino/input.ex:11]: - no docs or specs, do you have credo enabled?
- why is audio the default?

File: examples/video_and_audio.livemd [hunk 4]
```diff
@@ -18,6 +18,7 @@ Mix.install([
 
 ```elixir
 kino = Membrane.Kino.Player.new(:both)
```
Comment (mat-hek) [inline examples/video_and_audio.livemd:20]: both is quite ambiguous, maybe `[:audio, :video]`?

File: lib/membrane/kino/input.ex [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+      {:reply, :ok, assign(ctx, client: from)}
+    end
+  end
+
+  @impl true
+  def handle_call({:unregister, from}, _sender, ctx) do
+    if ctx.assigns.client == from do
+      {:reply, :ok, assign(ctx, client: nil)}
+    else
+      if ctx.assign.client == nil do
```
Comment (mat-hek) [inline lib/membrane/kino/input.ex:74]: what's `ctx.assign`?

File: lib/membrane/kino/input_source.ex [hunk 6]
```diff
@@ -0,0 +1,79 @@
+defmodule Membrane.Kino.InputSource do
```
Comment (mat-hek) [inline lib/membrane/kino/input_source.ex:1]: maybe just `Kino.Source`?

File: lib/membrane/kino/input_source.ex [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    RemoteStream,
+    Kino,
+    Time,
+    Buffer
+  }
+
+  def_options kino: [
+                spec: Kino.JS.Live.t(),
+                description: "Kino.JS.Live handle"
+              ]
```
Comment (mat-hek) [inline lib/membrane/kino/input_source.ex:20]: - doesn't it support only `Membrane.Kino.Input`?
- it should create and render an instance of `Membrane.Kino.Input` when it's not provided

File: lib/membrane/kino/input_source.ex [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+  end
+
+  @impl true
+  def handle_info({:audio_frame, info, binary}, ctx, state) do
+    duration = Map.get(info, "duration", 0)
+
+    buffer = %Buffer{
+      payload: binary,
+      metadata: %{
+        duration: Time.milliseconds(duration)
```
Comment (mat-hek) [inline lib/membrane/kino/input_source.ex:64]: do we need that?

File: lib/membrane/kino/player.ex [hunk 9]
```diff
@@ -34,24 +34,28 @@ defmodule Membrane.Kino.Player do
...
-  @jmuxer_check_interval_ms 1000
-
   @doc """
   Creates a new Membrane.Kino.Player component. Returns a handle to the player.
   Should be invoked at the end of the cell or explicitly rendered.
   """
   @spec new(player_type_t, []) :: t()
-  def new(type \\ :video, _opts \\ []) do
-    Kino.JS.Live.new(__MODULE__, type)
+  def new(type \\ :video, opts \\ []) do
```
Comment (mat-hek) [inline lib/membrane/kino/player.ex:48]: again, why is video the default?

File: test/decode_microphone_test.exs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  }
+
+  setup _ctx do
+    %{}
+  end
+
+  @tag :tmp_dir
+  @tag timeout: :infinity
+  test "Checks if audio and video are interleaved correctly", %{tmp_dir: tmp_dir} do
+    # Logger.configure(level: :error)
```
Comment (mat-hek) [inline test/decode_microphone_test.exs:24]:
```suggestion
```

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: lib/membrane/kino/input.ex (hunk 5 above)
Comment (Janix4000) [reply lib/membrane/kino/input.ex:72]: Typo

File: lib/membrane/kino/input.ex (hunk 1 above)
Comment (Janix4000) [reply lib/membrane/kino/input.ex:6]: Official Kino moudules follow this convention, see https://github.com/livebook-dev/kino/tree/main/lib/assets

File: lib/membrane/kino/input_source.ex (hunk 8 above)
Comment (Janix4000) [reply lib/membrane/kino/input_source.ex:64]: Not sure, but we can keep it for a potential future compatibility

File: lib/membrane/kino/input_source.ex (hunk 6 above)
Comment (Janix4000) [reply lib/membrane/kino/input_source.ex:1]: `Kino.Input.Source`? It follows convention `Kino.Player.Sink`

File: lib/membrane/kino/input_source.ex (hunk 8 above)
Comment (mat-hek) [reply lib/membrane/kino/input_source.ex]: what is future compatibility? 🤔

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: lib/membrane/kino/player.ex [hunk 11]
```diff
@@ -129,24 +129,24 @@ defmodule Membrane.Kino.Player do
...
-  @impl true
-  def handle_cast({:buffer, buffer, info}, ctx) when ctx.assigns.type in [:audio, :video] do
-    payload = {:binary, info, buffer}
+    if buffers[:audio] && not type[:audio] do
+      raise JMuxerError,
+        message: "Player was created without audio support, audio buffer provided"
+    end
+
+    video = Map.get(buffers, :video) || <<>>
+    audio = Map.get(buffers, :audio) || <<>>
```
Comment (mat-hek) [inline lib/membrane/kino/player.ex:146]:
```suggestion
    video = Map.get(buffers, :video, <<>>)
    audio = Map.get(buffers, :audio, <<>>)
```

File: lib/membrane/kino/player.ex [hunk 12]
```diff
@@ -129,24 +129,24 @@ defmodule Membrane.Kino.Player do
...
+
+    info = info |> Map.put(:video_size, byte_size(video)) |> Map.put_new(:type, :both)
```
Comment (mat-hek) [inline lib/membrane/kino/player.ex:148]: why is the `type` always `:both`?

File: lib/membrane/kino/player/player_sink.ex [hunk 13]
```diff
@@ -218,7 +209,8 @@ defmodule Membrane.Kino.Player.Sink do
   def handle_write({_mod, pad, _ref}, %Buffer{payload: payload}, _ctx, state) do
```
Comment (mat-hek) [inline lib/membrane/kino/player/player_sink.ex:209]:
```suggestion
  def handle_write(Pad.ref(pad, _id), %Buffer{payload: payload}, _ctx, state) do
```

File: lib/membrane/kino/player/player_sink.ex [hunk 14]
```diff
@@ -100,23 +97,17 @@ defmodule Membrane.Kino.Player.Sink do
...
     type = KinoPlayer.get_type(kino)
 
     tracks =
-      case type do
-        :video -> %{video: Track.new()}
-        :audio -> %{audio: Track.new()}
-        :both -> %{video: Track.new(), audio: Track.new()}
-      end
+      Enum.reduce(type, %{}, fn {type, exist}, acc ->
+        if exist do
```
Comment (mat-hek) [inline lib/membrane/kino/player/player_sink.ex:105]:
```suggestion
      Enum.reduce(type, %{}, fn {type, exists?}, acc ->
        if exists? do
```

File: lib/membrane/kino/input.ex [hunk 15]
```diff
@@ -122,20 +120,38 @@ defmodule Membrane.Kino.Input do
     if ctx.assigns.client do
       {:reply, {:error, :already_registered}, ctx}
     else
-      {:reply, :ok, assign(ctx, client: from)}
+      ref = Process.monitor(from)
```
Comment (mat-hek) [inline lib/membrane/kino/input.ex:123]: cool, do that in the player too

File: lib/membrane/kino/player.ex (hunk 12 above)
Comment (Janix4000) [reply lib/membrane/kino/player.ex:148]: It should be removed

File: lib/membrane/kino/input.ex (hunk 15 above)
Comment (Janix4000) [reply lib/membrane/kino/input.ex:123]: We decided to not do it right now

**Review by mat-hek — APPROVED** [review]

**Pull request merged**